### PR TITLE
test(mme): Add injection of state loaded in S1AP state manager

### PR DESCRIPTION
**Pull request by thhung**
Signed-off-by: Hung NGUYEN <nguyen.thanh.hung.91@gmail.com>

## Summary

- adding mocking methods to load predefined states on the S1AP task
- adding unittests related to the S1AP task using injected states  

Reference: #9607

## Test Plan

- make test_oai

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/c/core/oai/test/s1ap_task/data/s1ap_state_ATTACHED [hunk 1]
```diff
@@ -0,0 +1,3 @@
+
+K%G
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/test/s1ap_task/data/s1ap_state_ATTACHED]: Can we encode this file as binary (as we have for the UE state), we should be very strict on the data we are gonna save for mocking (for security purposes even if it's for testing)

File: lte/gateway/c/core/oai/test/s1ap_task/data/data_list.txt [hunk 2]
```diff
@@ -0,0 +1,2 @@
+IMSI001010000000001_S1AP_ATTACHED
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/test/s1ap_task/data/data_list.txt:1]: we should expand on the number of attached UE (both for SPGW and S1AP)

File: lte/gateway/c/core/oai/test/s1ap_task/mock_s1ap_op.cpp [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <fstream>
+#include <algorithm>
+
+#include "lte/gateway/c/core/oai/test/s1ap_task/mock_s1ap_op.h"
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/test/s1ap_task/mock_s1ap_op.cpp]: this should be the first included header

File: lte/gateway/c/core/oai/test/s1ap_task/mock_s1ap_op.cpp [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+// mocking the reading s1ap ue states from redis database by injecting local
+// samples
+status_code_e mock_read_s1ap_ue_state_db(
+    const std::vector<std::string>& ue_samples) {
+  hash_table_ts_t* state_ue_ht = get_s1ap_ue_state();
+
+  for (const auto& name_of_sample_file : ue_samples) {
+    UeDescription ue_proto = UeDescription();
+    std::fstream input(
+        name_of_sample_file.c_str(), std::ios::in | std::ios::binary);
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/test/s1ap_task/mock_s1ap_op.cpp]: what does `std::ios::in | std::ios::binary` mean here?

File: lte/gateway/c/core/oai/test/s1ap_task/mock_s1ap_op.cpp [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    if (!ue_proto.ParseFromIstream(&input)) {
+      std::cerr << "Failed to parse the sample: " << name_of_sample_file
+                << std::endl;
+      return RETURNerror;
+    }
+
+    ue_description_t* ue_context_p =
+        (ue_description_t*) calloc(1, sizeof(ue_description_t));
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/test/s1ap_task/mock_s1ap_op.cpp]: legit lint warning, let's use `reinterpret_cast`

**Review by ardzoht — CHANGES_REQUESTED** [review]

File: lte/gateway/c/core/oai/test/s1ap_task/mock_s1ap_op.cpp (hunk 4 above)
Comment (thhung) [reply lte/gateway/c/core/oai/test/s1ap_task/mock_s1ap_op.cpp]: It means we are reading a file in the binary mode. I will reorder these options to clarify.

File: lte/gateway/c/core/oai/test/s1ap_task/data/s1ap_state_ATTACHED (hunk 1 above)
Comment (thhung) [reply lte/gateway/c/core/oai/test/s1ap_task/data/s1ap_state_ATTACHED]: All of them (both UE states and this one) are encoded with the default encoder of protobuf. In cases that we cannot change the behavior of this encoder, we might need to encrypt the data after serializing to have an extra security layer.

File: lte/gateway/c/core/oai/test/s1ap_task/mock_s1ap_op.h [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+ * limitations under the License.
+ */
+#include <iostream>
+#include <fstream>
+#include <string>
+#include <vector>
+
+#include "lte/gateway/c/core/oai/common/common_defs.h"
+
+#define STR_DECODE_TABLE \
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/test/s1ap_task/mock_s1ap_op.h:21]: I would suggest adding a comment on top of these define statements to make it clear what is the usage for them (encoding s1ap task state)

File: lte/gateway/c/core/oai/test/s1ap_task/test_s1ap_mme_handlers_with_injected_state.cpp [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+                          INVALID_IMSI64, assoc_id, stream_id, 33, 10));
+
+  // State validation
+  ASSERT_TRUE(
+      is_enb_state_valid(state, assoc_id, S1AP_READY, number_attached_ue - 1));
+}
+
+TEST_F(S1apMmeHandlersWithInjectedStatesTest, HandleS1apPathSwitchRequest) {
+  ASSERT_EQ(task_zmq_ctx_main_s1ap_with_injected_states.ready, true);
+  ASSERT_TRUE(
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/test/s1ap_task/test_s1ap_mme_handlers_with_injected_state.cpp]: should we add asserts to make sure `number_attached_ue` is not changed before and after this test?

**Review by ardzoht — COMMENTED** [review]
left final set of comments, please resolve them and I will accept the PR, thanks!

File: lte/gateway/c/core/oai/test/s1ap_task/mock_s1ap_op.h [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+#define DEFAULT_S1AP_STATE_DATA_PATH \
+  "lte/gateway/c/core/oai/test/s1ap_task/data/s1ap_state_ATTACHED_encoded"
+
+std::vector<std::string> load_file_into_vector_of_line_content(
+    const std::string& data_folder_path, const std::string& file_name);
+status_code_e mock_read_s1ap_ue_state_db(
+    const std::vector<std::string>& ue_samples);
+status_code_e mock_read_s1ap_state_db(
+    const std::string& file_name_state_sample);
+void add_data_to_buffer(unsigned int& buf, int& buf_size, int data,
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/test/s1ap_task/mock_s1ap_op.h:48]: also add a comment here to describe the utility of these functions

**Review by ardzoht — APPROVED** [review]

**Comment by ardzoht** [comment]
@thhung the mme test job is failing on your latest pushed commit with:
```
[----------] 2 tests from S1apMmeHandlersWithInjectedStatesTest
[ RUN      ] S1apMmeHandlersWithInjectedStatesTest.GenerateUEContextReleaseCommand
/workspaces/magma/lte/gateway/c/core/oai/test/s1ap_task/test_s1ap_mme_handlers_with_injected_state.cpp:146: Failure
Value of: is_enb_state_valid(state, assoc_id, S1AP_READY, number_attached_ue - 1)
  Actual: false
Expected: true
```

**Comment by themarwhal** [comment]
@LKreutzer We should track Bazelifying this test. Filed https://github.com/magma/magma/issues/12102

**Comment by themarwhal** [comment]
Hello @thhung this unit test seems to have failed on my unrelated PR: https://github.com/magma/magma/runs/5544196751?check_suite_focus=true

Is it flaky?

**Comment by thhung** [comment]
Hi @themarwhal , thanks for your report. I will take a look.

**Comment by themarwhal** [comment]
@thhung It looks like it's been failing on master since this PR: https://github.com/magma/magma/actions/workflows/agw-workflow.yml?query=branch%3Amaster

**Comment by thhung** [comment]
@themarwhal I am trying to reproduce the error in my local setup. However, I have not been able to.  
```
      Start  7: test_itti
 7/25 Test  #7: test_itti ..................................   Passed    4.25 sec
      Start  8: test_amf_app
 8/25 Test  #8: test_amf_app ...............................   Passed    0.35 sec
      Start  9: test_smf_service_client
 9/25 Test  #9: test_smf_service_client ....................   Passed    0.04 sec
      Start 10: test_auth_service_client
10/25 Test #10: test_auth_service_client ...................   Passed    0.06 sec
      Start 11: test_s1ap
11/25 Test #11: test_s1ap ..................................   Passed   38.87 sec
      Start 12: test_nas_converter
12/25 Test #12: test_nas_converter .........................   Passed    0.07 sec
      Start 13: test_bstr
13/25 Test #13: test_bstr ..................................   Passed    0.01 sec
      Start 14: test_3gpp
14/25 Test #14: test_3gpp ..................................   Passed    0.04 sec
      Start 15: test_openflow_controller
15/25 Test #15: test_openflow_controller ...................   Passed    0.02 sec
      Start 16: test_imsi_encoder
16/25 Test #16: test_imsi_encoder ..........................   Passed    0.01 sec
``` 
The output is from the latest master.

**Comment by themarwhal** [comment]
> @themarwhal I am trying to reproduce the error in my local setup. However, I have not been able to.
> 
> ```
>       Start  7: test_itti
>  7/25 Test  #7: test_itti ..................................   Passed    4.25 sec
>       Start  8: test_amf_app
>  8/25 Test  #8: test_amf_app ...............................   Passed    0.35 sec
>       Start  9: test_smf_service_client
>  9/25 Test  #9: test_smf_service_client ....................   Passed    0.04 sec
>       Start 10: test_auth_service_client
> 10/25 Test #10: test_auth_service_client ...................   Passed    0.06 sec
>       Start 11: test_s1ap
> 11/25 Test #11: test_s1ap ..................................   Passed   38.87 sec
>       Start 12: test_nas_converter
> 12/25 Test #12: test_nas_converter .........................   Passed    0.07 sec
>       Start 13: test_bstr
> 13/25 Test #13: test_bstr ..................................   Passed    0.01 sec
>       Start 14: test_3gpp
> 14/25 Test #14: test_3gpp ..................................   Passed    0.04 sec
>       Start 15: test_openflow_controller
> 15/25 Test #15: test_openflow_controller ...................   Passed    0.02 sec
>       Start 16: test_imsi_encoder
> 16/25 Test #16: test_imsi_encoder ..........................   Passed    0.01 sec
> ```
> 
> The output is from the latest master.

@thhung  @ardzoht 
I see, that is weird... Do the LSAN outputs from the GitHub Action nodes seem legitimate? I think we should remove this test until we can get it to pass on CI. It will block all AGW related PRs from merging I think.

**Comment by ardzoht** [comment]
I agree @themarwhal I will put a revert PR since Hung is off today cc @thhung let's validate the flakiness locally first

**Comment by thhung** [comment]
Thanks @ardzoht . I will take a look right once I am back. @themarwhal Sorry for this inconvenience.

**Comment by themarwhal** [comment]
> Thanks @ardzoht . I will take a look right once I am back. @themarwhal Sorry for this inconvenience. 

Thanks @ardzoht and @thhung !

**Pull request merged**